File: src/main/java/net/vanillaoutsider/betterdogs/ai/PersonalityFollowOwnerGoal.java
// Copyright (C) 2026 Dasik (Rifaditya) | GNU GPLv3
// Verified against: Minecraft 26.3
package net.vanillaoutsider.betterdogs.ai;

import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.ai.goal.FollowOwnerGoal;
import net.minecraft.world.entity.animal.wolf.Wolf;
import net.vanillaoutsider.betterdogs.WolfExtensions;
import net.vanillaoutsider.betterdogs.WolfPersonality;
import net.vanillaoutsider.betterdogs.registry.BetterDogsGameRules;

public class PersonalityFollowOwnerGoal extends FollowOwnerGoal {
    private final Wolf wolf;
    private final double baseSpeedModifier;

    public PersonalityFollowOwnerGoal(Wolf wolf, double speedModifier, float minDistance, float maxDistance) {
        super(wolf, speedModifier, minDistance, maxDistance);
        this.wolf = wolf;
        this.baseSpeedModifier = speedModifier;
    }

    public PersonalityFollowOwnerGoal(Wolf wolf, double speedModifier, boolean leavesAllowed) {
        super(wolf, speedModifier, 10.0f, 2.0f);
        this.wolf = wolf;
        this.baseSpeedModifier = speedModifier;
    }

    private float betterdogs$followerSpacingOffset = 0.0f;
    private int spacingThrottleTimer = 0;
    private int recalcTimer = 0;

    public static class FollowerSpacingCache {
        private static final java.util.Map<java.util.UUID, CacheEntry> cache = new java.util.concurrent.ConcurrentHashMap<>();

        public static class CacheEntry {
            public final int followerCount;
            public final long expiryTime;

            public CacheEntry(int followerCount, long expiryTime) {
                this.followerCount = followerCount;
                this.expiryTime = expiryTime;
            }
        }

        public static int getCount(java.util.UUID ownerUuid, long currentTime) {
            CacheEntry entry = cache.get(ownerUuid);
            if (entry != null && currentTime < entry.expiryTime) {
                return entry.followerCount;
            }
            return -1;
        }

        public static int getLastKnownCount(java.util.UUID ownerUuid) {
            CacheEntry entry = cache.get(ownerUuid);
            return entry != null ? entry.followerCount : 0;
        }

        public static void update(java.util.UUID ownerUuid, int count, long currentTime, int lifetime) {
            cache.put(ownerUuid, new CacheEntry(count, currentTime + lifetime));
        }
    }

    private void updateFollowerSpacing(LivingEntity owner) {
        if (owner == null || this.wolf.level() == null) {
            this.betterdogs$followerSpacingOffset = 0.0f;
            return;
        }

        java.util.UUID ownerUuid = owner.getUUID();
        long currentTime = this.wolf.level().getGameTime();
        int count = FollowerSpacingCache.getCount(ownerUuid, currentTime);

        if (count < 0) {
            var nearbyWolves = this.wolf.level().getEntitiesOfClass(
                Wolf.class,
                owner.getBoundingBox().inflate(16.0),
                w -> w.isTame() && w.isOwnedBy(owner) && !w.isOrderedToSit()
            );
            count = nearbyWolves.size();
            FollowerSpacingCache.update(ownerUuid, count, currentTime, 40);
        }

        if (count <= 1) {
            this.betterdogs$followerSpacingOffset = 0.0f;
        } else {
            float multiplier = BetterDogsGameRules.getInt(
                this.wolf.level(),
                BetterDogsGameRules.BD_TAMED_PACK_SPREAD_MULTIPLIER,
                100
            ) / 100.0f;
            float maxExtra = BetterDogsGameRules.getInt(
                this.wolf.level(),
                BetterDogsGameRules.BD_TAMED_PACK_SPREAD_MAX,
                50
            ) / 10.0f;
            this.betterdogs$followerSpacingOffset = Math.min((float) Math.sqrt(count - 1) * multiplier, maxExtra);
        }
    }

    public float getStartDistance() {
        float baseDist = 10.0f;
        if (this.wolf instanceof WolfExtensions ext) {
            WolfPersonality personality = ext.betterdogs$getPersonality();
            baseDist = switch (personality) {
                case AGGRESSIVE -> (float) BetterDogsGameRules.getInt(this.wolf.level(), BetterDogsGameRules.BD_AGGRO_FOLLOW_START, 50);
                case PACIFIST -> (float) BetterDogsGameRules.getInt(this.wolf.level(), BetterDogsGameRules.BD_PACI_FOLLOW_START, 5);
                case NORMAL -> (float) BetterDogsGameRules.getInt(this.wolf.level(), BetterDogsGameRules.BD_NORMAL_FOLLOW_START, 10);
            };
        }
        return baseDist + this.betterdogs$followerSpacingOffset;
    }

    public float getStopDistance() {
        return 2.0f + this.betterdogs$followerSpacingOffset;
    }

    public float getTeleportThreshold() {
        float startDist = getStartDistance();
        if (startDist > 16.0f) {
            return 32.0f;
        }
        return startDist * 2.0f;
    }

    @Override
    public boolean canUse() {
        LivingEntity owner = this.wolf.getOwner();
        if (owner == null || owner.isSpectator() || this.wolf.isOrderedToSit() || this.wolf.isLeashed()) {
            return false;
        }
        if (this.wolf instanceof WolfExtensions ext && ext.betterdogs$isGuardMode()) {
            return false;
        }
        updateFollowerSpacing(owner);
        float startDist = getStartDistance();
        return this.wolf.distanceToSqr(owner) >= (startDist * startDist);
    }

    @Override
    public boolean canContinueToUse() {
        if (this.wolf.getNavigation().isDone() || this.wolf.isOrderedToSit() || this.wolf.isLeashed()) {
            return false;
        }
        if (this.wolf instanceof WolfExtensions ext && ext.betterdogs$isGuardMode()) {
            return false;
        }
        LivingEntity owner = this.wolf.getOwner();
        if (owner == null) {
            return false;
        }
        float stopDist = getStopDistance();
        return this.wolf.distanceToSqr(owner) > (stopDist * stopDist);
    }

    @Override
    public void start() {
        this.recalcTimer = 0;
    }

    @Override
    public void stop() {
        this.wolf.getNavigation().stop();
        super.stop();
    }

    @Override
    public void tick() {
        LivingEntity owner = this.wolf.getOwner();
        if (owner == null) {
            return;
        }

        if (++this.spacingThrottleTimer % 20 == 0) {
            updateFollowerSpacing(owner);
        }

        this.wolf.getLookControl().setLookAt(owner, 10.0f, (float) this.wolf.getMaxHeadXRot());

        if (!this.wolf.isLeashed() && !this.wolf.isPassenger()) {
            float teleportThreshold = getTeleportThreshold();
            if (this.wolf.distanceToSqr(owner) >= (teleportThreshold * teleportThreshold)) {
                teleportToOwner(owner);
                return;
            }
        }

        if (--this.recalcTimer <= 0) {
            this.recalcTimer = 10;
            double speed = this.baseSpeedModifier;
            if (this.wolf instanceof WolfExtensions ext) {
                WolfPersonality personality = ext.betterdogs$getPersonality();
                if (personality == WolfPersonality.AGGRESSIVE) {
                    speed *= 1.2;
                }
            }
            speed = net.vanillaoutsider.betterdogs.util.WolfCatchupHelper.calculateCatchupSpeed(this.wolf, owner, speed);
            this.wolf.getNavigation().moveTo(owner, speed);
        }
    }

    private void teleportToOwner(LivingEntity owner) {
        for (int i = 0; i < 10; ++i) {
            double dx = (this.wolf.getRandom().nextFloat() - 0.5) * 6.0;
            double dy = (this.wolf.getRandom().nextFloat() - 0.5) * 2.0;
            double dz = (this.wolf.getRandom().nextFloat() - 0.5) * 6.0;
            if (this.wolf.randomTeleport(owner.getX() + dx, owner.getY() + dy, owner.getZ() + dz, false)) {
                break;
            }
        }
    }
}

File: src/main/kotlin/net/vanillaoutsider/betterdogs/ai/PersonalityFollowOwnerGoal.kt
package net.vanillaoutsider.betterdogs.ai

import net.minecraft.world.entity.ai.goal.FollowOwnerGoal
import net.minecraft.world.entity.animal.wolf.Wolf
import net.vanillaoutsider.betterdogs.WolfPersonality
import net.vanillaoutsider.betterdogs.asExtended
import net.vanillaoutsider.betterdogs.config.BetterDogsConfig

/**
 * A custom FollowOwnerGoal that adjusts its start/stop distance based on the Wolf's personality.
 * This allows "Granular" control via the config.
 */
class PersonalityFollowOwnerGoal(
    private val wolf: Wolf,
    private val speedModifier: Double,
    private val leavesAllowed: Boolean
) : FollowOwnerGoal(wolf, speedModifier, 10.0f, 2.0f) {

    private var recalcTimer = 0

    // Helper to get config based on personality
    private fun getStartDistance(): Float {
        if (!wolf.isTame) return 10.0f // Fallback
        
        val ext = wolf.asExtended()
        return try {
             when (ext.`betterdogs$getPersonality`()) {
                WolfPersonality.AGGRESSIVE -> BetterDogsConfig.get().aggressiveFollowStart
                WolfPersonality.PACIFIST -> BetterDogsConfig.get().pacifistFollowStart
                WolfPersonality.NORMAL -> BetterDogsConfig.get().normalFollowStart
                else -> 10.0f
            }
        } catch (e: Exception) {
            10.0f
        }
    }

    private fun getStopDistance(): Float {
        if (!wolf.isTame) return 2.0f
        
        val ext = wolf.asExtended()
        return try {
            when (ext.`betterdogs$getPersonality`()) {
                WolfPersonality.AGGRESSIVE -> BetterDogsConfig.get().aggressiveFollowStop
                WolfPersonality.PACIFIST -> BetterDogsConfig.get().pacifistFollowStop
                WolfPersonality.NORMAL -> BetterDogsConfig.get().normalFollowStop
                else -> 2.0f
            }
        } catch (e: Exception) {
            2.0f
        }
    }

    /**
     * We override tick to hijack the logic.
     * Since FollowOwnerGoal's fields (startDistance, stopDistance) are private/final,
     * we cannot easily change them.
     * 
     * However, the main logic is in canUse() and canContinueToUse() and tick().
     * We can reimplement the checks using our dynamic getters.
     */
     
    // We override canUse to use our dynamic start distance
    override fun canUse(): Boolean {
        // Base checks from vanilla FollowOwnerGoal
        val owner = wolf.owner ?: return false
        if (owner.isSpectator) return false
        
        // Use OUR dynamic start distance
        if (wolf.distanceToSqr(owner) < (getStartDistance() * getStartDistance())) {
            return false
        }
        
        // Respect sitting/leashed
        if (wolf.isOrderedToSit || wolf.isLeashed) return false
        
        return true
    }

    // We override canContinueToUse to use our dynamic stop distance
    override fun canContinueToUse(): Boolean {
        if (wolf.navigation.isDone) return false
        if (wolf.isOrderedToSit || wolf.isLeashed) return false
        
        val owner = wolf.owner ?: return false
        return wolf.distanceToSqr(owner) > (getStopDistance() * getStopDistance())
    }

    override fun start() {
        super.start()
    }

    override fun stop() {
        wolf.navigation.stop()
        super.stop()
    }

    override fun tick() {
        // Teleport Logic (usually done in tick)
        // Vanilla: if dist > startDist (or 12), teleport.
        // We use getStartDistance() as the threshold for consistency,
        // or we can just call super.tick() IF we can trust it.
        // BUT super.tick() uses the hardcoded '10.0f' we passed in constructor.
        
        // Wait, super.tick() calls helper methods. 
        // Let's reimplement tick to be safe and use our values.
        
        val owner = wolf.owner ?: return
        wolf.lookControl.setLookAt(owner, 10.0f, wolf.maxHeadXRot.toFloat())
        
        if (--this.recalcTimer <= 0) {
            this.recalcTimer = 10
            
            // Move to owner
            if (!wolf.isLeashed && !wolf.isPassenger) {
                 if (wolf.distanceToSqr(owner) >= (getStartDistance() * getStartDistance())) {
                     wolf.navigation.moveTo(owner, speedModifier)
                 } else {
                     wolf.navigation.stop()
                 }
            }
        }
        
        // Teleport check
        // If distance is significantly larger than start distance (e.g. +2 blocks or just start distance)
        // Vanilla constant is usually 12 (144 sqr).
        // Let's use startDistance + 2 as teleport range? 
        // The user config says "Follow Start / Teleport Dist". So we treat them as same or close.
        // Let's say if dist > startDistance, we TRY to walk.
        // If dist > startDistance * 2 (or a hard cap), we teleport?
        // Actually, vanilla FollowOwnerGoal has `teleportToOwner()` logic.
        // We should just call the internal teleport logic if we are broken.
        
        // Simplified: Call super.tick() creates movement based on the 10.0 we passed in constructor.
        // WE CANNOT rely on super.tick() because it uses the old values.
        // We must fully implement movement here.
        
        if (!wolf.isLeashed && !wolf.isPassenger) {
             // Teleport if very far
             if (wolf.distanceToSqr(owner) >= 144.0 || wolf.distanceToSqr(owner) >= (getStartDistance() * getStartDistance() * 2)) {
                 // Custom teleport logic since super.teleportToOwner is private/inaccessible
                 for (i in 0..9) {
                     val dx = (wolf.random.nextFloat() - 0.5) * 4.0
                     val dy = (wolf.random.nextFloat() - 0.5) * 4.0
                     val dz = (wolf.random.nextFloat() - 0.5) * 4.0
                     if (wolf.randomTeleport(owner.x + dx, owner.y + dy, owner.z + dz, false)) break
                 }
             }
        }
    }
}
